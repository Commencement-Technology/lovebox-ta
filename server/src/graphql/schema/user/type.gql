type User @cacheControl(maxAge: 50){
  email: EmailAddress!
  password: String!
  joinDate: String! @cacheControl(maxAge: 0)
  username: String!
  id: ID!
}

type Query {
  users: [User]
  user(email: EmailAddress!): User
  me: User
}

input userDTO {
  email: EmailAddress!
  password: String!
  username: String!
}

type Mutation {
  register(user: userDTO): Boolean
  login(email: EmailAddress!, password: String!): String!
}
