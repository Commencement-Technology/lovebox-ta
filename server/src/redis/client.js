import { createClient } from 'redis';

export const redisClient = createClient({
    socket: {
        port: parseInt(process.env.REDIS_PORT) | 6379,
        host: process.env.REDIS_HOST | 'localhost',
    },
});
