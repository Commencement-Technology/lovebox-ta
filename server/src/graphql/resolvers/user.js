import { createUser, getUser, getAllUsers } from '../../redis/user';
import { v4 as uuidv4 } from 'uuid';
import { createToken, hashPassword, verifyPassword } from '../../helpers';
import consola from 'consola';
import {
    UserInputError,
    ForbiddenError,
    ValidationError,
} from 'apollo-server-express';

/**
 * User resolvers
 */

export const UserResolvers = {
    Query: {
        users: async (_, args) => {
            try {
                const users = await getAllUsers();
                return Object.values(users).map(user => JSON.parse(user));
            } catch (error) {
                consola.error(error);
                return [];
            }
        },
        user: async (_, { email }) => {
            try {
                const user = await getUser(email);
                return JSON.parse(user);
            } catch (error) {
                consola.error(error);
                return null;
            }
        },
        me: async (_, args, { user }) => {
            if (!user) {
                throw new ForbiddenError('You are not connected');
            }
            try {
                const me = await getUser(user.email);
                return JSON.parse(me);
            } catch (error) {
                throw error;
            }
        },
    },

    Mutation: {
        register: async (_, { user }) => {
            const { password, email } = user;

            try {
                const userInDB = await getUser(email);
                if (userInDB) {
                    throw new ForbiddenError('This user already exists');
                }
                const new_user = {
                    ...user,
                    password: await hashPassword(password),
                    id: uuidv4(),
                    joinDate: new Date(),
                };
                await createUser(new_user);
                return true;
            } catch (error) {
                throw error;
            }
        },

        login: async (_, { email, password }) => {
            if (!email || !password) {
                throw new UserInputError('Please provide your credentials');
            }
            try {
                const user = JSON.parse(await getUser(email));
                if (!user) {
                    throw new ForbiddenError('This user does not exist');
                }
                if (!(await verifyPassword(user.password, password))) {
                    throw new ValidationError('Your password is incorrect');
                }
                return createToken(email);
            } catch (error) {
                throw error;
            }
        },
    },
};
