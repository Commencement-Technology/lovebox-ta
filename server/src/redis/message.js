import { redisClient } from './client';

export const addMessage = async message =>
    await redisClient.hSet('messages', message.id, JSON.stringify(message));

export const getMessage = async id => await redisClient.hGet('messages', id);

export const deleteMessage = async id => await redisClient.hDel('messages', id);

export const getAllMessages = async () => await redisClient.hGetAll('messages');
