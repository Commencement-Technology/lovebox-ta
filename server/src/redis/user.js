import { redisClient } from './client';

export const createUser = async user => {
    await redisClient.hSet('users', user.email, JSON.stringify(user));
};

export const getUser = async email => await redisClient.hGet('users', email);

export const deleteUser = async email => await redisClient.hDel('users', email);

export const getAllUsers = async () => await redisClient.hGetAll('users');
