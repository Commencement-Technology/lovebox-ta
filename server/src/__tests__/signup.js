const { gql } = require('apollo-server');
const { createTestClient } = require('apollo-server-testing');
const { ApolloServer, schema, redisClient } = require('../server');
import consola from 'consola';
import { deleteUser } from '../redis/user';

const server = new ApolloServer({ ...schema, uploads: false });
redisClient
    .connect()
    .then(() =>
        consola.ready({
            message: 'Redis client is connected successfully',
            badge: true,
        })
    )
    .catch(error =>
        consola.error({
            message: `An error occured when connecting Redis client ${error}`,
            badge: true,
        })
    );
const SIGN_UP = gql`
    mutation {
        register(
            user: {
                email: "mama@lovebox.love"
                password: "1234567"
                username: "Pakenfit"
            }
        )
    }
`;

beforeAll(async () => await deleteUser('mama@lovebox.love'));
describe('User signup', () => {
    it('Should register user and return true', async () => {
        const { query } = createTestClient(server);
        const res = await query({ query: SIGN_UP });
        expect(res.errors).toBeUndefined();
        expect(res.data).toBeDefined();
        expect(res.data.register).toBeTruthy();
        expect(res).toMatchSnapshot();
    });
});
