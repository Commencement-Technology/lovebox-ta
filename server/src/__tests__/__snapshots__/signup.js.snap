// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User signup Should register user and return true 1`] = `
Object {
  "data": Object {
    "register": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
